monkey/agent_plugins/exploiters/log4shell/src/log4shell_exploit_client.py

```diff
@@ -0,0 +1,216 @@
+    ) -> Tuple[bool, bool]:
+        propagation_success = False
+        exploitation_success = self._assess_exploitation(
+            target_host,
+            exploit_class_downloaded,
+            options.exploit_download_timeout,
+            timestamp,
+            interrupt,
+        )
+
+        if exploitation_success:
+            propagation_success = self._assess_propagation(
+                target_host,
+                agent_binary_downloaded,
+                options.agent_binary_download_timeout,
+                interrupt,
+            )
+        return exploitation_success, propagation_success
+
+    def _assess_exploitation(
+        self,
+        target_host: TargetHost,
+        exploit_class_downloaded: Event,
+        exploit_download_timeout: float,
+        timestamp: float,
+        interrupt: Event,
+    ) -> bool:
+        exploitation_success = Log4ShellExploitClient._wait_for_victim_to_download_java_bytecode(
+            exploit_class_downloaded, exploit_download_timeout, interrupt
+        )
+        if exploitation_success:
+            self._publish_exploitation_event(target_host, timestamp, True)
+        else:
+            error_message = "Timed out while waiting for victim to download the java bytecode"
+            logger.debug(error_message)
+            self._publish_exploitation_event(
+                target_host, timestamp, False, error_message=error_message
+            )
+
+        return exploitation_success
+
+    @staticmethod
+    def _wait_for_victim_to_download_java_bytecode(
+        exploit_class_downloaded: Event, exploit_download_timeout: float, interrupt: Event
+    ) -> bool:
+        return Log4ShellExploitClient._wait_for_event(
+            exploit_class_downloaded, exploit_download_timeout, interrupt
+        )
+
+    def _assess_propagation(
+        self,
+        target_host: TargetHost,
+        agent_binary_downloaded: Event,
+        agent_binary_download_timeout: float,
+        interrupt: Event,
+    ) -> bool:
+        propagation_success = Log4ShellExploitClient._wait_for_victim_to_download_agent(
+            agent_binary_downloaded, agent_binary_download_timeout, interrupt
+        )
+        self._publish_propagation_event(target_host, success=propagation_success)
+
+        return propagation_success
+
+    @staticmethod
+    def _wait_for_victim_to_download_agent(
+        agent_binary_downloaded: Event, agent_binary_download_timeout: float, interrupt: Event
+    ) -> bool:
+        return Log4ShellExploitClient._wait_for_event(
+            agent_binary_downloaded, agent_binary_download_timeout, interrupt
+        )
+
+    @staticmethod
+    def _wait_for_event(event: Event, timeout: float, interrupt: Event) -> bool:
+        timer = EggTimer()
+        timer.set(timeout)
+
+        while not timer.is_expired() and not interrupt.is_set():
+            if event.wait(VICTIM_WAIT_SLEEP_TIME_SEC):
+                return True
+
+        return event.is_set()
```

Comment:
The `EggTimer` logic is pointless here.
```suggestion
        if event.wait(timeout):
                return True
                
        return False
```

Reply:
Could also just do:
```suggestion
        return event.wait(timeout)
```

Reply:
https://docs.python.org/3/library/threading.html#threading.Event.wait

Reply:
Nope we couldn't. In this way we wait for a fraction but no more then the timeout. This way we are more responsive and safe. For example what if we wait for the timeout and the event happens afterwards?

Reply:
> For example what if we wait for the timeout and the event happens afterwards?

I don't understand what you mean.

Reply:
Never mind. We're doing this because of the interrupt.